fdk.go

```diff
@@ -3,94 +3,142 @@ package fdk
+func WithContext(ctx context.Context, fnctx Context) context.Context {
+	return context.WithValue(ctx, ctxKey, fnctx)
+}
+
+type key struct{}
+
+var ctxKey = new(key)
+
+// Context contains all configuration for a function invocation
+type Context interface {
+	// Config is a map of all env vars set on a function, the base set of fn
+	// headers in addition to any app and function configuration.
+	Config() map[string]string
+
+	// Header are the headers sent to this function invocation
+	Header() http.Header
+
+	// ContentType is Header().Get("Content-Type") but with 15 less chars, you are welcome
+	// XXX(discuss): is this necessary? Header.Get("Content-Type") is easy enough for a caveman...
+	ContentType() string
+
+	// CallID is the call id for this function invocation
+	CallID() string
+
+	// TODO should we unwind all known env var fields as methods here, as well?
+}
+
+// HTTPContext contains all configuration for a function invocation sourced
+// from an http gateway trigger, which will make the function appear to receive
+// from the client request they were sourced from, with no additional headers.
+type HTTPContext interface {
+	Context
+
+	// RequestURL is the request url from the gateway client http request
+	RequestURL() string
+
+	// RequestMethod is the request method from the gateway client http request
+	RequestMethod() string
+}
+
+type baseCtx struct {
+	header http.Header
+	config map[string]string
+	callID string
 }
 
-// Ctx provides access to Config and Headers from fn.
-type Ctx struct {
-	// Header are the unmodified headers as sent to the container, see
-	// HTTPHeader for specific trigger headers
-	Header http.Header
-	// HTTPHeader are the request headers as they appear on the original HTTP request,
-	// for an http trigger.
-	HTTPHeader http.Header
-	Config     map[string]string
-	RequestURL string
-	Method     string
+type httpCtx struct {
+	// XXX(reed): if we embed we won't preserve the original headers. since we have an
+	// interface handy now we could change this under the covers when/if we want... idk
+	baseCtx
+	requestURL    string
+	requestMethod string
+}
+
+func (c baseCtx) Config() map[string]string { return c.config }
+func (c baseCtx) Header() http.Header       { return c.header }
+func (c baseCtx) CallID() string            { return c.callID }
+func (c baseCtx) ContentType() string       { return c.header.Get("Content-Type") }
+
+func (c httpCtx) RequestURL() string    { return c.requestURL }
+func (c httpCtx) RequestMethod() string { return c.requestMethod }
+
+func ctxWithDeadline(ctx context.Context, fnDeadline string) (context.Context, context.CancelFunc) {
+	t, err := time.Parse(time.RFC3339, fnDeadline)
+	if err == nil {
+		return context.WithDeadline(ctx, t)
+	}
+	return context.WithCancel(ctx)
 }
 
-// AddHeader will add a header on the function response, for hot function
-// formats.
+// AddHeader will add a header onto the function response
 func AddHeader(out io.Writer, key, value string) {
```

Comment:
In other places I have ways to set the HTTP headers for a trigger using the same sort of contract (in other FDKs the context is mutable though which is not very golang) - and also some helpers to set the status code header value in response

Reply:
not entirely clear here, I think this answers your question though, the way this is all plumbed is that users will call this and then later on after their handler is done we'll wrap it up http gateway style. we could put it on the context, this is odd since output isn't an http.ResponseWriter but it is an http.ResponseWriter, am just not sure what to do with it. I could see how having things on context makes them more flexible towards other trigger types down the road, it's not clear what 'feels' better, and i'm just terrible at feeling things...